core/frontend/src/IModelApp.ts

```diff
@@ -763,6 +763,16 @@ export class IModelApp {
     return this.localization.getLocalizedString(`iModelJs:${key.scope}.${key.val}`, key);
   }
 
+  /**
+   * Resets the formatsProvider back to the default [[BasicFormatsProvider]].
+   * @beta
+   */
+  public static resetFormatsProvider() {
+    const oldFormatsProvider = this.formatsProvider;
+    this.formatsProvider = new BasicFormatsProvider();
+    oldFormatsProvider.onFormatsChanged.raiseEvent({ formatsChanged: "all"})
+  }
```

Comment:
This showcased an interesting situation.

Let's say property widget, which relies on formatting, registers an `onFormatsChanged` listener. Then, at some point, application using the widget decides to change or reset the formats provider - widget would be listening on the stale provider. How do we solve that?

One idea I have is to always have the same formats provider on `IModelApp` - it would be used to re-raise events from and forward calls to underlying implementation of the actual formats provider - the "basic" one or app-supplied. So instead of setting them on the `IModelApp`, you'd set them on the IModelApp's formats provider. What do you think?

Reply:
I see that yes... without modifying the interface of a formatsProvider further to tackle this, I think the root of the solution is having some sort of 'manager' in IModelApp to handle this. Maybe we can repurpose the QuantityFormatter to do so: add a setFormatsProvider(), a BeEvent for onFormatsProvider changed (or onFormatsChanged that raises the FormatsChangedArg defined in core-quantity). We keep the formatsProvider opt in IModelAppOptions, and if user supplied one then we set quantityFormatter.formatsProvider during startup.

Users would get formats through IModelApp.quantityFormatter.formatsProvider.

Reply:
I don't mind formats provider being on quantity formatter rather than `IModelApp`, but I'd hate to see `onFormatsProviderChanged` or another `onFormatsChanged` event.

- The former would mean that, in my property widget example, I have to listen to two events: `QuantityFormatter.onFormatsProviderChanged` and `FormatsProvider.onFormatsChanged`.
- The latter would duplicate the `onFormatsChanged` event and I think it wouldn't be clear which one to use for consumers.

My initial suggestion doesn't require changing the `FormatsProvider` interface, but instead make the implementation act as the manager that you mention. We don't need to expose the implementation - it's enough that `IModelApp` or `QuantityFormatter` knows it.